Add Blockgroups level fix #73

jbousquin opened the pull request:
Added "blockgroup" as level for census and ACS ([Issue #73](https://github.com/cenpy-devs/cenpy/issues/73)). Tiger layers are 16 (census) and 10 (ACS).

On the issue thread @dfolch pointed out 2 items for additional consideration:
1. cenpy API name for this level. cenpy uses "blockgroup" consistently, so I used the same except for geo_unit since census requires "block group" as the string. Looking to other similar project - ACS.R uses block.group for geo.make but both "blockgroup" and "block group" in documentation/messages. tidycensus (also R) uses "block group" as the expected geography value (accepts 'cbg' as well).
2. Missing years. 'block group' is a defined geo unit in [2013](https://api.census.gov/data/2013/acs/acs5/examples.html) on but not 2010-2012 (e.g. [2012](https://api.census.gov/data/2012/acs/acs5/examples.html)). There is already a NotImplementedError on ACS years before 2013. This will need to be revisited if earlier years are ever added.

I did some manual tests but wasn't sure how you wanted unit tests added (in test_functional_products.py?)

jbousquin committed:
Update products.py

Added {'blockgroup':10} to layer_lookup for tiger decennial map service layer

jbousquin committed:
Update products.py

Added query hierarchy for blockgroup

jbousquin committed:
Fix geo_unit for census

jbousquin committed:
Added blockgroup layer to ACS product

jbousquin committed:
Test blockgroup from decennial county & ACS

ljwolf commented:
Hey, thanks for this! this is excellent! 

The `test_functional_products.py` file is a direct conversion of the [demo notebook announcing the product API](https://nbviewer.jupyter.org/github/cenpy-devs/cenpy/blob/master/notebooks/product-api.ipynb), which should probably just use something like the pytest [nbval](https://github.com/computationalmodelling/nbval) plugin to actually ensure that the products don't change. 

So, if you add a query to the blockgroups to that code (as you have) that's fine!

jbousquin committed:
Update test_functional.py

Changed key to 2011 since there seems to be a problem on the census side with 2012 & 2013.

jbousquin commented:
Set up travis with my fork so I could re-run tests since some of the errors are inconsistent (e.g. HTTP Error 500: Internal Server Error). Let me know if there is a better way to document this.

The key error in test_functional.py is new in the past week, looks like a problem on the census side with 2012 & 2013. They have it listed under available APIs [link](https://www.census.gov/data/developers/data-sets/cbp-nonemp-zbp/cbp-api.2012.html) but [calls](https://api.census.gov/data/2012/cbp) give a 404 error. For now I just changed the key to CBP2011.

yonil7 commented:
Is this feature scheduled for the next release?

jbousquin committed:
Merge branch 'cenpy-devs:master' into master

jbousquin committed:
Merge branch 'cenpy-devs:master' into master

jbousquin committed:
Dynamic layer level lookup

Get alt levels for diff years (see PR: https://github.com/cenpy-devs/cenpy/pull/127/commits/bdfc399dcf29f6bcbf10cbc7738accd623800714)

jbousquin committed:
Redirect mapservice to census for ACS 2020

Catch and redirect if ACS 2020 to Census 2020. Could be done in tiger, but this way it will be assigned to attribute. Could be done earlier in product but suspect this will be more concise. Could also be done as a try/catch instead, but current key error handling is good.

jbousquin committed:
Update products.py

Update catch to allow 2021 as latest

jbousquin committed:
Update test_functional_products.py

Add 2020 and 2021 as test case

jbousquin committed:
Update test_functional_products.py

Revert

jbousquin committed:
Revert to test orginal

jbousquin committed:
Will test with use case since build test already fails

jbousquin committed:
Fix typo
